### PR TITLE
Fix goroutine leak

### PR DESCRIPTION
**Pull request by fgimenez**
Towards https://github.com/giantswarm/giantswarm/issues/3607

The goroutines used for launching `ForwardPorts` are leaking, this PR adds a context return from the goroutine as soon as the tunnel is ready, keeping the ability to check errors too.

### CODE REVIEW

**Review by xh3b4sd — CHANGES_REQUESTED** [review]
This issue calls for unit tests at least.

**Comment by fgimenez** [comment]
@xh3b4sd not sure if all the k8s connections involved would be easy to mock on a unit test, let me know if you have something in mind.

As an alternative I can write an e2e test, IMO it would be good to have them anyway in this repo, and use leaktest for making sure there are no leaks, WDYT?

**Comment by xh3b4sd** [comment]
Whatever test you find suitable makes sense I think.

File: tunnel.go [hunk 1]
```diff
@@ -108,7 +114,8 @@ type Tunnel struct {
 	TunnelConfig
 	Local int
 
-	stopChan chan struct{}
+	readyChan chan struct{}
+	stopChan  chan struct{}
```
Comment (kopiczko) [inline tunnel.go]: Why do we need readyChan to be a part of the `Tunnel`?

**Review by kopiczko — COMMENTED** [review]
Other than my comment LGTM. I'm big on not having redundant fields in structs because it brings confusion.

File: tunnel.go (hunk 1 above)
Comment (fgimenez) [reply tunnel.go]: I wanted to align `stopChan` and `readyChan`, but you are right in that the latter is only used in this method, no need to add it to the struct (I even have a vague memory of having discussed this before :D). Reverted, thx!

Comment (kopiczko) [reply tunnel.go]: 👍

**Review by kopiczko — APPROVED** [review]
LGTM. I don't know how to test goroutine leaks. But I'm happy to learn.

**Comment by xh3b4sd** [comment]
For a unit test I would have set up the routine management and create and delete the setup a couple of times while checking the programs goroutine number does not increase in an usual way. The `runtime` package provides all these information. See https://golang.org/pkg/runtime/#NumGoroutine.

**Comment by tuommaki** [comment]
> For a unit test I would have set up the routine management and create and delete the setup a couple of times while checking the programs goroutine number does not increase in an usual way.

If this method is successful, I'd say the number of create/delete cycles must be moderately large (hundreds) and then check with some 50-150 error marginal to account for changes in Go runtime.

**Comment by fgimenez** [comment]
I'm going for the e2e test that emulates better a production use case. Also using leaktest, I have some experinece with it for detecting this kind of problems, https://github.com/fortytw2/leaktest, better IMO than building our own solution. Will push shortly the changes.

**Comment by fgimenez** [comment]
Finally going with an unit test and leaktest. Running on an actual integration tests, leaktest uncovered more dangling goroutines, but those are related to how the actual port forwarding connection is done, their number doesn't grow over time.

Checked also that the test covers the fixed issue, without the changes in this PR it fails with:
```
$ go test ./...
E0717 14:10:35.981021    8277 portforward.go:178] lost connection to pod
--- FAIL: TestGoroutineLeak (5.00s)
	leaktest.go:129: leaktest: timed out checking goroutines
	leaktest.go:147: leaktest: leaked goroutine: goroutine 10 [chan send]:
		github.com/giantswarm/k8sportforward.newTunnel.func1(0xc4200ae5a0, 0xc4201d4000)
			/home/fgimenez/workspace/gocode/src/github.com/giantswarm/k8sportforward/tunnel.go:137 +0x5c
		created by github.com/giantswarm/k8sportforward.newTunnel
			/home/fgimenez/workspace/gocode/src/github.com/giantswarm/k8sportforward/tunnel.go:135 +0x31c
FAIL
FAIL	github.com/giantswarm/k8sportforward	5.012s
```

**Comment by fgimenez** [comment]
@xh3b4sd PTAL ^

File: tunnel_test.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+	}
+
+	streamConn, err := spdy.NewClientConnection(conn)
+	if err != nil {
+		t.Fatalf("could not create connection to server %v", err)
+	}
+	dialer := &fakeDialer{
+		conn: streamConn,
+	}
+	_, err = newTunnel(dialer, tunnelCfg)
```
Comment (xh3b4sd) [inline tunnel_test.go:58]: How is the tunnel closed after it is established and we verified what we want to test? How does the test finally end?

Comment (fgimenez) [reply tunnel_test.go:58]: With this setup the tunnel is closed after `ts` server is closed. 

Anyway we are not having issues with not closed tunnels, the point of the test is checking that this goroutine https://github.com/giantswarm/k8sportforward/blob/master/tunnel.go#L89-L91 is not leaked anymore, after the analysis with pprof described here https://github.com/giantswarm/giantswarm/issues/3607 it turned out that it is the source of the problem. 

As stated above, going back to the old goroutine definition makes the test fail.

Comment (xh3b4sd) [reply tunnel_test.go:58]: What makes the test server closing then?

Comment (fgimenez) [reply tunnel_test.go:58]: > What makes the test server closing then?

`defer ts.Close()` here https://github.com/giantswarm/k8sportforward/pull/7/files#diff-b247ac39dcfc87620ab8cb858fc0e663R34

Comment (xh3b4sd) [reply tunnel_test.go:58]: Oh I see. With you.

**Review by xh3b4sd — APPROVED** [review]

**Pull request merged**